crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4668,6 +4668,33 @@ impl<'db> TypeInferenceBuilder<'db> {
         let symbol_table = self.index.symbol_table(file_scope_id);
         let use_def = self.index.use_def_map(file_scope_id);
 
+        // Perform narrowing with applicable constraints between the current scope and the enclosing scope.
+        let narrow_with_applicable_constraints = |mut ty, enclosing_scope_id| {
+            let mut scope = scope;
+            loop {
```

Comment:
We already walk up enclosing scopes below. Is it possible that we can just do one walk, both looking for the symbol and collecting applicable bindings, instead of two walks? It feels like this will help us avoid logic duplication (e.g. for things like "names in class scopes are not visible to nested scopes" which will help us avoid bugs.